src/libsync/syncfilestatustracker.cpp

```diff
@@ -31,229 +40,143 @@ namespace OCC {
-            // cut the folder path
-            dirRec = _syncEngine->journal()->getFileRecord(parentPath);
-
-            isDir = true;
-        }
-        if( dirRec.isValid() && !dirRec._remotePerm.isNull()) {
-            if( (isDir && !dirRec._remotePerm.contains("K"))
-                    || (!isDir && !dirRec._remotePerm.contains("C")) ) {
-                status.set(SyncFileStatus::StatusError);
-            }
-        }
-    }
-    return status;
-}
+    // If we're not currently syncing that file, look it up in the database to know if it's shared
+    SyncJournalFileRecord rec = _syncEngine->journal()->getFileRecord(fileName);
+    if (rec.isValid())
+        return fileStatus(rec.toSyncFileItem());
 
-void SyncFileStatusTracker::slotThreadTreeWalkResult(const SyncFileItemVector& items)
-{
-    addErroredSyncItemPathsToList(items, &_stateLastSyncItemsWithErrorNew);
+    // Must be a new file, wait for the filesystem watcher to trigger a sync
+    return SyncFileStatus();
 }
 
 void SyncFileStatusTracker::slotAboutToPropagate(SyncFileItemVector& items)
 {
-    addErroredSyncItemPathsToList(items, &_stateLastSyncItemsWithErrorNew);
-}
+    std::set<Problem> oldProblems;
+    std::swap(_syncProblems, oldProblems);
 
-void SyncFileStatusTracker::slotSyncFinished()
-{
-    _stateLastSyncItemsWithError = _stateLastSyncItemsWithErrorNew;
-    _stateLastSyncItemsWithErrorNew.clear();
+    foreach (const SyncFileItemPtr &item, items) {
+        qDebug() << Q_FUNC_INFO << "Investigating" << item->destination() << item->_status;
+
+        if (showErrorInSocketApi(*item))
+            _syncProblems.insert({item->_file, SyncFileStatus::StatusError});
+        else if (showWarningInSocketApi(*item))
+            _syncProblems.insert({item->_file, SyncFileStatus::StatusWarning});
+
+        QString systemFileName = _syncEngine->localPath() + item->destination();
+        // the trailing slash for directories must be appended as the filenames coming in
+        // from the plugins have that too. Otherwise the matching entry item is not found
+        // in the plugin.
+        if( item->_type == SyncFileItem::Type::Directory )
+            systemFileName += QLatin1Char('/');
+        emit fileStatusChanged(systemFileName, fileStatus(*item));
+    }
+
+    // Make sure to push any status that might have been resolved indirectly since the last sync
+    // (like an error file being deleted from disk)
+    for (auto it = _syncProblems.begin(); it != _syncProblems.end(); ++it)
+        oldProblems.erase(*it);
+    for (auto it = oldProblems.begin(); it != oldProblems.end(); ++it) {
+        if (it->severity == SyncFileStatus::StatusError)
+            invalidateParentPaths(it->path);
+        emit fileStatusChanged(_syncEngine->localPath() + it->path, fileStatus(it->path));
+    }
 }
 
 void SyncFileStatusTracker::slotItemCompleted(const SyncFileItem &item)
 {
+    qDebug() << Q_FUNC_INFO << item.destination() << item._status;
```

Comment:
There are this, and a bunch of other, qDebug()s left in this patch - intentional?